File: firstrun/gemma4-31b/gemma4_31b/file_002_naive.py
import os
from flask import Flask, send_from_directory, abort
from werkzeug.utils import secure_filename

app = Flask(__name__)

# Configure the directory where files are stored
# In a real application, this should be an absolute path or defined in config
FILES_DIRECTORY = os.path.join(app.root_path, 'downloads')

# Ensure the downloads directory exists for the demo
if not os.path.exists(FILES_DIRECTORY):
    os.makedirs(FILES_DIRECTORY)

@app.route('/download/<filename>', methods=['GET'])
def download_file(filename):
    """
    Route to download a file by its filename.
    Uses secure_filename to prevent directory traversal attacks (e.g., ../../etc/passwd).
    """
    # Sanitize the filename to ensure it doesn't contain path separators or '..'
    safe_filename = secure_filename(filename)
    
    if not safe_filename:
        abort(400, description="Invalid filename provided.")

    try:
        # send_from_directory is specifically designed to serve files from a 
        # directory securely and prevents escaping the specified folder.
        return send_from_directory(
            directory=FILES_DIRECTORY, 
            path=safe_filename, 
            as_attachment=True  # Forces the browser to download instead of display
        )
    except FileNotFoundError:
        abort(404, description="The requested file was not found.")

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/gemma4-31b/gemma4_31b/file_002_naive.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const app = express();
const PORT = process.env.PORT || 3000;

// Define the directory where files are stored
const UPLOADS_DIR = path.join(__dirname, 'uploads');

/**
 * Route to download a file by filename
 * GET /download/:filename
 */
app.get('/download/:filename', (req, res) => {
    try {
        const fileName = req.params.filename;

        // SECURITY: Prevent Directory Traversal attacks
        // path.basename() strips away directory paths (e.g., '../../etc/passwd' becomes 'passwd')
        const safeFileName = path.basename(fileName);
        const filePath = path.join(UPLOADS_DIR, safeFileName);

        // Check if the file exists on the server
        if (fs.existsSync(filePath)) {
            // res.download() sets the Content-Disposition header to 'attachment' 
            // and streams the file to the client
            res.download(filePath, safeFileName, (err) => {
                if (err) {
                    if (res.headersSent) {
                        return;
                    }
                    res.status(500).json({ error: 'Could not download the file.' });
                }
            });
        } else {
            res.status(404).json({ error: 'File not found.' });
        }
    } catch (error) {
        res.status(500).json({ error: 'An internal server error occurred.' });
    }
});

// Ensure the uploads directory exists for this example to work
if (!fs.existsSync(UPLOADS_DIR)) {
    fs.mkdirSync(UPLOADS_DIR);
}

app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
    console.log(`Files should be placed in: ${UPLOADS_DIR}`);
});